### DISCUSSION TITLE
Translations in .po files instead of PHP arrays

### DISCUSSION THREAD

**Discussion by angrybrad** [craftcms/cms#9165] [Ideas]
> Created by: Francis Bourassa (francis@ascense.ca) on 2016/06/13 18:36:52 +0000
> Votes at time of UserVoice import: 7

-----------------------------------

The usage of php arrays for static translations, especially plugins, makes the work of translators very tedious. The way translations work, even the slightest change in a translated string will invalidate it, that's normal. In Craft CMS though, there is no way to see a list of strings that have been invalidated and need revison. You have to find out by yourself by browsing the templates (correct me if I'm wrong). 

The workflow for PO files is very simple for translators:
-Generate a .pot file (generaly CLI tool, could be in the control panel)
-Update or generate a .po file for the language, from the .pot file.
-Use a Po file editor, see strings that need revision. 

You can even upload you .po file to google translate, get suggestions, etc. It's widely used so there is a lot of good tools out there. 
PHP arrays are hard to read and offers no functionality. They are prone to syntax errors and duplicates.

Craft CMS is used globally and could benefits more users around the word by using this standard, right now it's simply lacking.

**Comment by angrybrad** [comment]
> Posted by Wesley B (wesleyb@phmgmt.com) on 2016/09/01 16:40:23 +0000
-----------------------------------

There is also another post about supporting PO files: #1173

I would also like to see support for `gettext` / `.po`; this would allow automation of translations using a standard. Here are three companies that work with `.po` files from digital sources (like app and websites):

- CrowdIn: https://crowdin.com/
- LingoTek: https://lingotek.com/
- SpeakLike: http://www.speaklike.com/

**Comment by Mosnar** [comment]
I've always found the php array style translations a bit odd and has raised some eyebrows with translators in the past.

Yii 2 natively supports Gettext style translation with `po` and `mo` files and can be used by configuring your app.php file as such:

```php
'components' => [
            'i18n' => [
                'translations' => [
                    'site' => [
                        'class' => yii\i18n\GettextMessageSource::class,
                        'sourceLanguage' => 'en-US',
                        'catalog' => 'site',
                        'basePath' => '@translations',
                        'forceTranslation' => true,
                    ]
                ]
            ],
],
```

There are a couple caveats...
1. Each translation requires an extra item to denote the context as `site` for some reason. So a full translation might look like this:
```po
msgctxt "site"
msgid "Hello!"
msgstr "Bonjour!"
```
2. I built a rudimentary wrapper in Craft around the [Twig-Gettext-Extractor](https://github.com/umpirsky/Twig-Gettext-Extractor) package, but ended up not having much luck. I haven't yet tried POEdit Pro's native Twig support.

In retrospect, it might be better to look into using the DbMessageSource or building a custom MessageSource in Craft that's more congruent with its practices; however, I wanted to contribute the code above in case anyone needs it.
